File: domains/finance/projects/stock-market/development/software/ios/StockPrediction/Models/SpeakerSuggestions.swift
import Foundation

// MARK: - Contextual speaker suggestions
//
// Maps a ticker (and, where useful, sector keywords) to the most relevant of the
// pre-seeded YVIS speakers to follow for that stock. Used to render a subtle
// "Track for this stock: <name>" hint on the stock detail screen.
// Single source of truth — names match the seeded speakers in VideoIntelligenceView.

enum SpeakerSuggestions {

    /// Speakers most relevant to a specific ticker.
    private static let byTicker: [String: [String]] = [
        "TSLA":    ["Elon Musk", "Cathie Wood"],
        "MSTR":    ["Michael Saylor"],
        "BTC-USD": ["Michael Saylor"],
        "ETH-USD": ["Michael Saylor"],
        "COIN":    ["Michael Saylor", "Cathie Wood"],
        "NVDA":    ["Jensen Huang"],
        "AAPL":    ["Tim Cook"],
        "ROKU":    ["Cathie Wood"],
        "RBLX":    ["Cathie Wood"],
        "HOOD":    ["Cathie Wood"],
        "SQ":      ["Cathie Wood"],
        "PATH":    ["Cathie Wood"],
        "BRK-B":   ["Warren Buffett"],
        "BRK-A":   ["Warren Buffett"],
    ]

    /// Sector / keyword fallbacks when the ticker isn't explicitly mapped.
    private static let bySectorKeyword: [(keyword: String, speakers: [String])] = [
        ("ev",             ["Elon Musk"]),
        ("auto",           ["Elon Musk"]),
        ("crypto",         ["Michael Saylor"]),
        ("blockchain",     ["Michael Saylor"]),
        ("semiconductor",  ["Jensen Huang"]),
        ("semis",          ["Jensen Huang"]),
        ("artificial intelligence", ["Jensen Huang"]),
        ("technology",     ["Jensen Huang"]),
        ("innovation",     ["Cathie Wood"]),
        ("disruptive",     ["Cathie Wood"]),
        ("financial",      ["Warren Buffett"]),
        ("bank",           ["Jerome Powell", "Warren Buffett"]),
    ]

    /// Returns suggested speaker names for the given ticker / optional sector,
    /// most relevant first. Empty when nothing relevant is found.
    static func speakers(forTicker ticker: String, sector: String? = nil) -> [String] {
        let key = ticker.uppercased()
        if let direct = byTicker[key] { return direct }

        // Crypto pairs not individually listed → Saylor.
        if key.hasSuffix("-USD") { return ["Michael Saylor"] }

        if let sector, !sector.isEmpty {
            let lower = sector.lowercased()
            for entry in bySectorKeyword where lower.contains(entry.keyword) {
                return entry.speakers
            }
        }
        return []
    }
}

File: domains/finance/projects/stock-market/development/software/android/core/src/main/java/com/prediction/stockmarket/ui/stock/SpeakerSuggestions.kt
package com.prediction.stockmarket.ui.stock

// Contextual speaker suggestions
//
// Maps a ticker (and, where useful, sector keywords) to the most relevant of the
// pre-seeded YVIS speakers to follow for that stock. Used to render a subtle
// "Track for this stock: <name>" hint on the stock detail screen.
// Single source of truth — names match the seeded speakers in VideoIntelligenceManager.

object SpeakerSuggestions {

    private val byTicker: Map<String, List<String>> = mapOf(
        "TSLA" to listOf("Elon Musk", "Cathie Wood"),
        "MSTR" to listOf("Michael Saylor"),
        "BTC-USD" to listOf("Michael Saylor"),
        "ETH-USD" to listOf("Michael Saylor"),
        "COIN" to listOf("Michael Saylor", "Cathie Wood"),
        "NVDA" to listOf("Jensen Huang"),
        "AAPL" to listOf("Tim Cook"),
        "ROKU" to listOf("Cathie Wood"),
        "RBLX" to listOf("Cathie Wood"),
        "HOOD" to listOf("Cathie Wood"),
        "SQ" to listOf("Cathie Wood"),
        "PATH" to listOf("Cathie Wood"),
        "BRK-B" to listOf("Warren Buffett"),
        "BRK-A" to listOf("Warren Buffett"),
    )

    private val bySectorKeyword: List<Pair<String, List<String>>> = listOf(
        "ev" to listOf("Elon Musk"),
        "auto" to listOf("Elon Musk"),
        "crypto" to listOf("Michael Saylor"),
        "blockchain" to listOf("Michael Saylor"),
        "semiconductor" to listOf("Jensen Huang"),
        "semis" to listOf("Jensen Huang"),
        "artificial intelligence" to listOf("Jensen Huang"),
        "technology" to listOf("Jensen Huang"),
        "innovation" to listOf("Cathie Wood"),
        "disruptive" to listOf("Cathie Wood"),
        "financial" to listOf("Warren Buffett"),
        "bank" to listOf("Jerome Powell", "Warren Buffett"),
    )

    /**
     * Suggested speaker names for the given ticker / optional sector, most
     * relevant first. Empty when nothing relevant is found.
     */
    fun speakers(ticker: String, sector: String? = null): List<String> {
        val key = ticker.uppercase()
        byTicker[key]?.let { return it }

        // Crypto pairs not individually listed → Saylor.
        if (key.endsWith("-USD")) return listOf("Michael Saylor")

        if (!sector.isNullOrEmpty()) {
            val lower = sector.lowercase()
            for ((keyword, speakers) in bySectorKeyword) {
                if (lower.contains(keyword)) return speakers
            }
        }
        return emptyList()
    }
}
